src/odemis/driver/tmcm.py

```diff
@@ -2502,3 +2553,871 @@ def _parseMessage(self, msg):
+                # receives updates both values are already updated.
+                self._updatePosition(axes)
+                # read-only so manually notify
+                self.referenced.notify(self.referenced.value)
+                # # Update the global variable, based on the referenced axes
+                # self._update_ref()
+
+    def homing(self, future, timeout=30, set_new_home=True):
+        """
+        Function to execute the configured Homing Method on the node.
+        This function does exactly the same as the corresponding function of canopen.BaseNode402 except that
+        it is compatible with the cancelling procedure, especially when cancelling very early (this is not
+        possible with the original function).
+        :param CancellableFuture future: the future for the referencing thread
+        :param int timeout: Timeout value (default: 30)
+        :param bool set_new_home: Defines if the node should set the home offset
+        object (0x607C) to the current position after the homing procedure (default: true)
+        :return bool: If the homing was complete with success
+        """
+        # Use init lock to make sure we're not cancelling during the state changes.
+        with future._init_lock:
+            if future._must_stop.is_set():
+                # Cancelled before homing started (init_lock acquired here after cancelling function)
+                return False
+            previus_op_mode = self._node.op_mode
+            self._node.state = 'SWITCHED ON'
+            self._node.op_mode = 'HOMING'
+            # The homing process will initialize at operation enabled
+            self._node.state = 'OPERATION ENABLED'
+            homingstatus = 'IN PROGRESS'
+            self._node.controlword = State402.CW_OPERATION_ENABLED | Homing.CW_START
+        t = time.time() + timeout
+        try:
+            while homingstatus not in ('TARGET REACHED', 'ATTAINED'):
+                for key, value in Homing.STATES.items():
+                    # check if the value after applying the bitmask (value[0])
+                    # corresponds with the value[1] to determine the current status
+                    bitmaskvalue = self._node.statusword & value[0]
+                    if bitmaskvalue == value[1]:
+                        homingstatus = key
+                if homingstatus in ('INTERRUPTED', 'ERROR VELOCITY IS NOT ZERO', 'ERROR VELOCITY IS ZERO'):
+                    raise RuntimeError('Unable to home. Reason: {0}'.format(homingstatus))
+                time.sleep(0.001)
+                if time.time() > t:
+                    raise RuntimeError('Unable to home, timeout reached')
+            if set_new_home:
+                actual_position = self._node.sdo[ACTUAL_POSITION].raw
+                self._node.sdo[HOME_OFFSET].raw = actual_position  # home offset (0x607C)
+                logging.info('Homing offset set to {0}'.format(actual_position))
```

Comment:
I have the feeling this works only for the referencing of the first axis. Or is there something special with referencing which make all the axes move at the same time? For now, we can't test with several axes, so unless it's clear to you how to deal with multiple axes, I'd suggest to put a big TODO at the top of this function. Maybe even pass an "axis" parameter and raise an Error if axis != 0.